### ISSUE TITLE
Investigate small current oscillations

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/capacitor-lab-basics#125]
Found in the event stream, there is a very small oscillation in current.  It is not clear why this is there, and should be investigated.

The observed oscillation went from 0 to  0 to -1.49191642661101e-14 back to 0.

**Comment by andrewadare** [comment]
This looks similar to the [ESPB energy non-conservation issue](https://github.com/phetsims/energy-skate-park-basics/issues/355): the model is computing a quantity using an approximation, and the PhET-iO data stream exposes imperfections that are not otherwise visible.

In AbstractCircuit.js, I = dQ/dt in updateCurrentAmplitude is a good place to begin investigating.

**Comment by andrewadare** [comment]
I took some time this afternoon to save the phet-io datastream and make some plots. Here's currentAmplitudeProperty vs time:
![currentamplitudeproperty](https://cloud.githubusercontent.com/assets/154163/18530810/720ac296-7a90-11e6-8b54-1a7a46c17a4a.png)

and here's platesVoltageProperty vs time, divided by the (unrealistically large) light bulb resistance:

![platevoltageproperty](https://cloud.githubusercontent.com/assets/154163/18530838/9d4fb5ce-7a90-11e6-8146-7541e8ba1cdc.png)

Apart from having slightly different lifetimes (the two quantities are set to zero at slightly different times at the end of their decay), they are basically identical.

I left out the final value of zero from the phet-io output, so the points could be plotted on a semi-log scale. I can't see any evidence of small-current oscillations here.

@jessegreenberg @samreid wasn't this issue originally brought up by the UBC colleagues? As you see, I'm having trouble reproducing it, and I can't find that discussion.

**Comment by samreid** [comment]
I'm not sure where this report originated.  Perhaps this is just the current spike when battery voltage changes?

**Comment by andrewadare** [comment]
If this is happening near the end of the decay, there are additional things changing in the sim due to truncations put in place so the current doesn't asymptotically tail off to zero forever. There's some discussion relevant to this in #58.

**Comment by jessegreenberg** [comment]
I just forwarded @andrewadare the original email thread that reported this issue, but this might be the relevant information:

>It appears the currentAmplitudeProperty change is driven from the model step function and not an immediate (synchronous) result of the battery voltage change.

>capacitorLabBasics.lightBulbScreen.view.lightBulbCircuitNode.batteryNode.sliderNode.thumbInputListener dragged {"x":57,"y":342}
    86 capacitorLabBasics.lightBulbScreen.model.circuit.battery.voltageProperty changed {"oldValue":-0.4231084421979052,"newValue":-0.47012049133100575}
        87 capacitorLabBasics.lightBulbScreen.model.circuit.switchedCapacitor.platesVoltageProperty changed {"oldValue":-0.4231084421979052,"newValue":-0.47012049133100575}
            88 capacitorLabBasics.lightBulbScreen.model.plateChargeMeter.valueProperty changed {"oldValue":-3.746202147220253e-14,"newValue":-4.1624468302447246e-14}
            89 capacitorLabBasics.lightBulbScreen.model.storedEnergyMeter.valueProperty changed {"oldValue":7.925248773344044e-15,"newValue":9.784257744869188e-15}
90 capacitorLabBasics.lightBulbScreen.model.circuit.currentAmplitudeProperty changed {"oldValue":0,"newValue":-1.49191642661101e-14}
91 capacitorLabBasics.lightBulbScreen.model.circuit.currentAmplitudeProperty changed {"oldValue":-1.49191642661101e-14,"newValue":0}

>This looks like a small oscillation from 0 to -1.49191642661101e-14 and back to zero, which I am not sure of the physical cause of.

>In the preceding lines, we also see a battery voltage change, then a subsequent current change (2 events, with a drag sneaking in between them), and the 2nd event taking the current amplitude to 0.

>capacitorLabBasics.lightBulbScreen.view.lightBulbCircuitNode.batteryNode.sliderNode.thumbInputListener dragged {"x":56,"y":341}
    78 capacitorLabBasics.lightBulbScreen.model.circuit.battery.voltageProperty changed {"oldValue":-0.3760963930648047,"newValue":-0.4231084421979052}
        79 capacitorLabBasics.lightBulbScreen.model.circuit.switchedCapacitor.platesVoltageProperty changed {"oldValue":-0.3760963930648047,"newValue":-0.4231084421979052}
            80 capacitorLabBasics.lightBulbScreen.model.plateChargeMeter.valueProperty changed {"oldValue":-3.32995746419578e-14,"newValue":-3.746202147220253e-14}
            81 capacitorLabBasics.lightBulbScreen.model.storedEnergyMeter.valueProperty changed {"oldValue":6.2619249567162826e-15,"newValue":7.925248773344044e-15}
82 capacitorLabBasics.lightBulbScreen.model.circuit.currentAmplitudeProperty changed {"oldValue":-1.435326493187836e-13,"newValue":-5.4057751042139283e-14}
83 capacitorLabBasics.lightBulbScreen.view.lightBulbCircuitNode.batteryNode.sliderNode.thumbInputListener dragged {"x":57,"y":341}
84 capacitorLabBasics.lightBulbScreen.model.circuit.currentAmplitudeProperty changed {"oldValue":-5.4057751042139283e-14,"newValue":0}

**Comment by andrewadare** [comment]
> I'm not sure where this report originated.

@samreid This report originated from you :) The mail referenced in https://github.com/phetsims/capacitor-lab-basics/issues/125#issuecomment-247393285 is yours. In fact you already identified the cause of this issue, namely that current and voltage are updated on different schedules:

> It appears the currentAmplitudeProperty change is driven from the model step function and not an immediate (synchronous) result of the battery voltage change.

While troubleshooting #130 I confirmed that this is true, and it was the cause of that bug in a corner case.

@jessegreenberg filed this issue to investigate the phenomenon. Indeed, it's been investigated by multiple people, and the cause is clear. The circuit state updating behavior would not be easy to change without major rewriting and lots of side-effects. We'd need to see if there's a performance hit with e.g. adding more update function calls to step(dt). For all this, the payoff is questionable.

As a side-note, Sarah Perez didn't identify this as a problem at all--it was found by @samreid incidentally and reported in the middle of an email thread, then logged by @jessegreenberg to look at.

Can I get some additional opinions on what to do at this point?

**Comment by samreid** [comment]
Perhaps the oscillation is physically correct?  For instance, I=dq/dt and Q=VC.  Hence if the voltage or capacitance change, it would cause a small temporary current.  I'm not too familiar with the sim, can @ariel-phet @jessegreenberg or @andrewadare check my physics?

**Comment by andrewadare** [comment]
In a pure RC circuit like this, there should be no oscillation (except for contrivances like wiggling the thumb slider). There are transient currents, but they should always decay exponentially (thus monotonically), both on charge and on discharge of the capacitor.

Do you mean that we may be seeing a tiny spike, followed by a decay that only lasts about one timestep? Although that's not strictly un-physical, it's more plausible that we're just seeing the updates get slightly out of sync sometimes.

**Comment by samreid** [comment]
Yes, add `console.log(this.currentAmplitude);` to updateCurrentAmplitude and try a few things in the sim.  They are tiny transient spikes when the charge, voltage or capacitance change.

**Comment by andrewadare** [comment]
Here's `currentAmplitudeProperty` distilled from the phet-io output while dragging the battery slider up at a slow-to-medium speed to charge the capacitor:
```
97 {"oldValue":0,"newValue":2.665378245606214e-13}
98 {"oldValue":2.665378245606214e-13,"newValue":0}
104 {"oldValue":0,"newValue":1.4678894685946436e-13}
105 {"oldValue":1.4678894685946436e-13,"newValue":0}
111 {"oldValue":0,"newValue":4.823065396811381e-13}
112 {"oldValue":4.823065396811381e-13,"newValue":0}
119 {"oldValue":0,"newValue":1.4067274074031994e-13}
120 {"oldValue":1.4067274074031994e-13,"newValue":0}
126 {"oldValue":0,"newValue":3.5538376608082423e-13}
127 {"oldValue":3.5538376608082423e-13,"newValue":0}
133 {"oldValue":0,"newValue":1.0550455555524863e-13}
134 {"oldValue":1.0550455555524863e-13,"newValue":0}
140 {"oldValue":0,"newValue":2.4115326984056964e-13}
141 {"oldValue":2.4115326984056964e-13,"newValue":0}
147 {"oldValue":0,"newValue":1.7769188304040414e-13}
148 {"oldValue":1.7769188304040414e-13,"newValue":0}
154 {"oldValue":0,"newValue":1.1641881992303296e-13}
155 {"oldValue":1.1641881992303296e-13,"newValue":0}
161 {"oldValue":0,"newValue":1.6880728888838392e-13}
167 {"oldValue":1.6880728888838392e-13,"newValue":1.35045831110719e-13}
173 {"oldValue":1.35045831110719e-13,"newValue":1.0550455555523956e-13}
174 {"oldValue":1.0550455555523956e-13,"newValue":0}
180 {"oldValue":0,"newValue":1.5346117171672616e-13}
182 {"oldValue":1.5346117171672616e-13,"newValue":0}
188 {"oldValue":0,"newValue":4.2201822222097716e-13}
194 {"oldValue":4.2201822222097716e-13,"newValue":1.7769188304042007e-13}
195 {"oldValue":1.7769188304042007e-13,"newValue":0}
201 {"oldValue":0,"newValue":1.6076884656036624e-13}
202 {"oldValue":1.6076884656036624e-13,"newValue":0}
208 {"oldValue":0,"newValue":1.1253819259226479e-13}
209 {"oldValue":1.1253819259226479e-13,"newValue":0}
215 {"oldValue":0,"newValue":1.9859681045693932e-13}
221 {"oldValue":1.9859681045693932e-13,"newValue":1.6880728888838519e-13}
227 {"oldValue":1.6880728888838519e-13,"newValue":2.1100911111049568e-13}
233 {"oldValue":2.1100911111049568e-13,"newValue":2.110091111104815e-13}
234 {"oldValue":2.110091111104815e-13,"newValue":0}
240 {"oldValue":0,"newValue":1.7769188304042007e-13}
241 {"oldValue":1.7769188304042007e-13,"newValue":0}
247 {"oldValue":0,"newValue":2.597035213667552e-13}
249 {"oldValue":2.597035213667552e-13,"newValue":0}
255 {"oldValue":0,"newValue":1.2504243621362849e-13}
261 {"oldValue":1.2504243621362849e-13,"newValue":1.985968104569215e-13}
267 {"oldValue":1.985968104569215e-13,"newValue":5.957904313708009e-13}
273 {"oldValue":5.957904313708009e-13,"newValue":1.9859681045693932e-13}
274 {"oldValue":1.9859681045693932e-13,"newValue":0}
280 {"oldValue":0,"newValue":2.2507638518451106e-13}
286 {"oldValue":2.2507638518451106e-13,"newValue":1.9859681045693932e-13}
292 {"oldValue":1.9859681045693932e-13,"newValue":2.9357789371894063e-13}
298 {"oldValue":2.9357789371894063e-13,"newValue":6.707733335611204e-14}
299 {"oldValue":6.707733335611204e-14,"newValue":0}
```
The zeros occur because the current charge (specifically the value returned by `this.getTotalCharge()`) happens to have the same value as `this.previousTotalCharge` in a given timestep (so dQ/dt = 0).

As you can see, if you drag at the right speed, you get something that looks like the reported "oscillation" between zero and a finite current. To a sim user, there's no problem; to a phet-io datastream analyzer, naively collecting `oldValue` or `newValue` for direct use would not work well.

This issue is labeled as a bug. Is it a bug, or just how the sim works?

**Comment by andrewadare** [comment]
If this is a bug (and I'm not sure whether it is), then it's a bug in the way the sim was designed, and there will be a fair amount of effort required to make it work differently.

**Comment by ariel-phet** [comment]
I think we define "bugs" as something obvious to the user that either appears broken or depicts incorrect physics/behavior.

Since there is nothing obvious to the user, this is not a bug

**Comment by andrewadare** [comment]
@ariel-phet I agree. The sim works just as it was intended to, but phet-io puts a microscope on the details of how something is calculated internally (just like the ESPB [energy reporting issue](https://github.com/phetsims/energy-skate-park-basics/issues/355)). Perhaps whatever strategy gets decided there will have some bearing on what to do here.

**Comment by jessegreenberg** [comment]
Removing assignment since this issue has been deferred.

**Comment by pixelzoom** [comment]
@jessegreenberg  Does the "deferred" status of this issue mean that it does not need to be addressed for 1.0.0 release?

**Comment by pixelzoom** [comment]
@jessegreenberg  Does the "deferred" status of this issue mean that it does not need to be addressed for 1.0.0 release?

**Comment by jessegreenberg** [comment]
My understanding was this is a has no impact on the sim behavior, but is only apparent with PhET-iO features.

If this needs to be addressed for PhET-iO, then deferred seems appropriate and does not need to be adressed for a 1.0.0 release.  @ariel-phet said
> Since there is nothing obvious to the user, this is not a bug

So perhaps this can be closed.

**Comment by jessegreenberg** [comment]
I noticed @phet-steele added a 👎. This is a phet-io problem and I honestly don't know if it needs to be addressed for the 1.0.0 release. @ariel-phet do you know?

**Comment by samreid** [comment]
In my opinion, the main reason for this issue was to make sure the model didn't have latent bugs or flaws in the implementation.  If the model is working as desired, then I don't see any problem with this fluctuation in the PhET-iO API (except that we may have to document or explain what is happening to curious clients).

**Comment by ariel-phet** [comment]
I imagine @phet-steele was saying "don't close" 

@pixelzoom this does not need to be addressed for 1.0.0 release

**Comment by jessegreenberg** [comment]
Thanks @samreid, https://github.com/phetsims/capacitor-lab-basics/issues/125#issuecomment-250007033 isolated the cause of this, and https://github.com/phetsims/capacitor-lab-basics/issues/125#issuecomment-294913842 made this sound like a non-issue for phet-io. @phet-steele I am going to add some documentation for the phet-io instance and then close, sound good?
